src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -903,6 +903,23 @@ ucs_status_t uct_dc_mlx5_ep_fc_pure_grant_send(uct_dc_mlx5_ep_t *ep,
     ucs_assert((sizeof(uint8_t) + sizeof(sender_ep)) <=
                UCT_IB_MLX5_AV_FULL_SIZE);
 
+    if (iface->flags & UCT_DC_MLX5_IFACE_FLAG_FC_EP_FAILED) {
+        /* If the EP has FAILED flag set, it means that the DCI has already been
+         * assigned to the FC endpoint and it has not been reset yet. It is
+         * currently waiting for all operation compeltions as a part of handling
+         * failure. Add the operation to the pending queue for furhter
+         * processing after error recovery. FC EP arbiter group mustn't be
+         * scheduled to avoid dispatching pending operations while DCI isn't
+         * recovered from the failure */
+        ucs_assertv(ep->dci != UCT_DC_MLX5_EP_NO_DCI, "fc_ep %p: dci %u", ep,
+                    ep->dci);
```

Comment:
seems redundant. or it should be before the if. should not even try to send if no DCI

Reply:
why redundant?
ep is allocated after this `if`.
here we just make sure that dci is allocated

Reply:
maybe move after this if?
if no dci, and we're not sending anyway now (return NO_RES), should not care if the ep is failed or not

Reply:
i mean, check if ep is failed only after allocating a dci

Reply:
i see - done